packages/beacon-state-transition/src/block/processExecutionPayload.ts

```diff
@@ -46,10 +47,19 @@ export function processExecutionPayload(
   // if executionEngine is null, executionEngine.onPayload MUST be called after running processBlock to get the
   // correct randao mix. Since executionEngine will be an async call in most cases it is called afterwards to keep
   // the state transition sync
-  if (executionEngine && !executionEngine.notifyNewPayload(payload)) {
+  if (
+    type === allForks.BlockType.Full &&
+    executionEngine &&
+    !executionEngine.notifyNewPayload(payload as bellatrix.ExecutionPayload)
+  ) {
     throw Error("Invalid execution payload");
   }
 
+  const transactionsRoot =
+    type === allForks.BlockType.Blinded
+      ? (payload as bellatrix.ExecutionPayloadHeader).transactionsRoot
+      : ssz.bellatrix.Transactions.hashTreeRoot((payload as bellatrix.ExecutionPayload).transactions);
```

Comment:
This can be solved without the type with type guards

```ts
function getTransactionsRoot(block: BlockFull | BlockBlinded): Root {
  if ((block as BlockBlinded).transactionsRoot) {
    return (block as BlockBlinded).transactionsRoot
  } else {
    // hashTreeRoot
  }
}

Reply:
changed something on these lines everywhere except assemble block :+1: